scylla-cql/src/frame/request/batch.rs

```diff
@@ -54,9 +54,21 @@ where
         // Serializing queries
         types::write_short(self.statements_count.try_into()?, buf);
 
-        for (statement_num, statement) in self.statements.clone().enumerate() {
+        let mut n_serialized_statements = 0usize;
+        let mut value_lists = self.values.batch_values_iter();
+        for statement in self.statements.clone() {
             statement.serialize(buf)?;
-            self.values.write_nth_to_request(statement_num, buf)?;
+            value_lists.write_next_to_request(buf).ok_or_else(|| {
+                ParseError::BadDataToSerialize(
+                    "Mismatch between statement counts for batch query".to_owned(),
+                )
+            })??;
+            n_serialized_statements += 1;
+        }
+        if n_serialized_statements != self.statements_count {
+            return Err(ParseError::BadDataToSerialize(
+                "Mismatch between statement counts for batch query".to_owned(),
+            ));
         }
```

Comment:
I'm not sure if I understand the purpose of this check. AFAIK `n_serialized_statements == self.statements_count` is always true because we only construct `Batch` internally and `self.statements_count` is just the length of the iterator (see `Connection::batch_with_consistency`). If that was the purpose then I don't think this check is needed.

Reply:
As Batch is constructed in a separate file and seems to be generic to allow for evolutions in the way statements are managed, it seems to me error-prone for the future to rely on something non-local to this function to guarantee that the serialization respects the scylla protocol (that is, we have serialized as many values as we have announced).
If that were to happen, it looks like it would lead to an extremely hard to debug error.

Reply:
added rationale comment